### PR TITLE
add graph_datacell and sparse_graph_datacell

### PR DESCRIPTION
**Pull request by LHT129**
issue: #40 
- graph_datacell for bottom graph
- sparse_graph_datacell for high level graph

### CODE REVIEW

File: src/data_cell/graph_datacell.h [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                  const IndexCommonParam& common_param);
+
+    void
+    InsertNeighborsById(InnerIdType id, const Vector<InnerIdType>& neighbor_ids) override;
+
+    uint32_t
+    GetNeighborSize(InnerIdType id) override;
+
+    void
+    GetNeighbors(InnerIdType id, Vector<InnerIdType>& neighbor_ids) override;
```
Comment (wxyucs) [inline src/data_cell/graph_datacell.h]: add const modifier if possible

File: src/data_cell/graph_datacell.h [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    void
+    Serialize(StreamWriter& writer) override;
+
+    void
+    Deserialize(StreamReader& reader) override;
+
+private:
+    std::shared_ptr<BasicIO<IOTmpl>> io_{nullptr};
+
+    uint32_t code_line_size_{0};
```
Comment (wxyucs) [inline src/data_cell/graph_datacell.h:82]: when does this variable will be assigned a value

File: src/data_cell/graph_datacell.h [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    this->code_line_size_ = this->maximum_degree_ * sizeof(InnerIdType) + sizeof(uint32_t);
+}
+
+template <typename IOTmpl>
+void
+GraphDataCell<IOTmpl, false>::InsertNeighborsById(InnerIdType id,
+                                                  const Vector<InnerIdType>& neighbor_ids) {
+    if (neighbor_ids.size() > this->maximum_degree_) {
+        logger::error(fmt::format(
+            "insert neighbors count {} more than {}", neighbor_ids.size(), this->maximum_degree_));
```
Comment (wxyucs) [inline src/data_cell/graph_datacell.h]: what is the next step after logging errors, return or continue?

File: src/data_cell/graph_datacell.h [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+template <typename IOTmpl>
+void
+GraphDataCell<IOTmpl, false>::Deserialize(StreamReader& reader) {
+    GraphInterface::Deserialize(reader);
+    this->io_->Deserialize(reader);
+    StreamReader::ReadObj(reader, this->code_line_size_);
+}
+
+}  // namespace vsag
```
Comment (wxyucs) [inline src/data_cell/graph_datacell.h]: add a blank line at the end of file

File: src/data_cell/graph_datacell_test.cpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+#include "catch2/catch_template_test_macros.hpp"
+#include "fmt/format-inl.h"
+#include "graph_interface_test.h"
+#include "io/io_headers.h"
+using namespace vsag;
+
+template <typename IOTemp>
+void
+TestGraphDataCell(const nlohmann::json& graph_param,
```
Comment (wxyucs) [inline src/data_cell/graph_datacell_test.cpp]: create an alias of `nlomann::json`, njson or some other name, avoiding third-party dependency in too many files

File: src/data_cell/graph_datacell_test.cpp [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    auto max_id = 10000;
+    for (auto count : counts) {
+        auto graph = std::make_shared<GraphDataCell<IOTemp, false>>(graph_param, io_param, param);
+        GraphInterfaceTest test(graph);
+        auto other = std::make_shared<GraphDataCell<IOTemp, false>>(graph_param, io_param, param);
+        test.BasicTest(max_id, count, other);
+    }
+}
+
+TEST_CASE("graph basic test [ut][graph datacell]") {
```
Comment (wxyucs) [inline src/data_cell/graph_datacell_test.cpp]: incorrect format, `TEST_CASE(<name_only_use_lowercase_letter_and_underscore_and_numbers>, "[tag1][tag2]")`

File: src/data_cell/graph_interface.cpp [hunk 7]
```diff
@@ -0,0 +1,52 @@
+// Copyright 2024-present the vsag project
```
Comment (wxyucs) [inline src/data_cell/graph_interface.cpp:2]: add a blank line before this

File: src/data_cell/graph_interface.cpp [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+#include "graph_interface.h"
+
+#include "graph_datacell.h"
+#include "io/io_headers.h"
+#include "sparse_graph_datacell.h"
+
+namespace vsag {
+
+std::shared_ptr<GraphInterface>
```
Comment (wxyucs) [inline src/data_cell/graph_interface.cpp]: add an alias of std::shared_ptr<GraphInterface>, and replace this line with it

File: src/data_cell/graph_interface_test.cpp [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        for (auto& [key, value] : maps) {
+            Vector<InnerIdType> neighbors(allocator.get());
+            other->GetNeighbors(key, neighbors);
+            REQUIRE(memcmp(neighbors.data(), value->data(), value->size() * sizeof(InnerIdType)) ==
+                    0);
+        }
+
+        infile.close();
+    }
+}
```
Comment (wxyucs) [inline src/data_cell/graph_interface_test.cpp]: add a blank line at the end of file

File: src/data_cell/graph_interface_test.h [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+public:
+    GraphInterfaceTest(std::shared_ptr<GraphInterface> graph) : graph_(graph){};
+
+    void
+    BasicTest(uint64_t max_id, uint64_t count, std::shared_ptr<GraphInterface> other);
+
+public:
+    std::shared_ptr<GraphInterface> graph_{nullptr};
+};
+}  // namespace vsag
```
Comment (wxyucs) [inline src/data_cell/graph_interface_test.h]: add a blank line at the end of file

File: src/data_cell/sparse_graph_datacell.cpp [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    uint64_t size;
+    StreamReader::ReadObj(reader, size);
+    for (auto i = 0; i < size; ++i) {
+        InnerIdType key;
+        StreamReader::ReadObj(reader, key);
+        this->neighbors_[key] = std::make_unique<vsag::Vector<InnerIdType>>(allocator_);
+        StreamReader::ReadVector(reader, *(this->neighbors_[key]));
+    }
+}
+}  // namespace vsag
```
Comment (wxyucs) [inline src/data_cell/sparse_graph_datacell.cpp]: add a blank line at the end of file

File: src/data_cell/sparse_graph_datacell.h [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    Deserialize(StreamReader& reader) override;
+
+private:
+    uint32_t code_line_size_{0};
+    Allocator* allocator_{nullptr};
+    UnorderedMap<InnerIdType, std::unique_ptr<Vector<InnerIdType>>> neighbors_;
+    std::shared_mutex neighbors_map_mutex_{};
+};
+
+}  // namespace vsag
```
Comment (wxyucs) [inline src/data_cell/sparse_graph_datacell.h]: add a blank line at the end of file

File: src/data_cell/sparse_graph_datacell.h [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+    void
+    Serialize(StreamWriter& writer) override;
+
+    void
+    Deserialize(StreamReader& reader) override;
+
+private:
+    uint32_t code_line_size_{0};
+    Allocator* allocator_{nullptr};
```
Comment (wxyucs) [inline src/data_cell/sparse_graph_datacell.h]: add const modifier

File: src/data_cell/sparse_graph_datacell.h [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    explicit SparseGraphDataCell(Allocator* allocator = nullptr, uint32_t max_degree = 32);
+
+    void
+    InsertNeighborsById(InnerIdType id, const Vector<InnerIdType>& neighbor_ids) override;
+
+    uint32_t
+    GetNeighborSize(InnerIdType id) override;
+
+    void
+    GetNeighbors(InnerIdType id, Vector<InnerIdType>& neighbor_ids) override;
```
Comment (wxyucs) [inline src/data_cell/sparse_graph_datacell.h]: add const modifier

File: src/data_cell/sparse_graph_datacell.cpp [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+    }
+}
+
+void
+SparseGraphDataCell::Deserialize(StreamReader& reader) {
+    GraphInterface::Deserialize(reader);
+    StreamReader::ReadObj(reader, this->code_line_size_);
+    uint64_t size;
+    StreamReader::ReadObj(reader, size);
+    for (auto i = 0; i < size; ++i) {
```
Comment (wxyucs) [inline src/data_cell/sparse_graph_datacell.cpp]: auto i = 0 -> uint64_t 0

File: src/data_cell/graph_interface_test.h [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+#include <algorithm>
+#include <random>
+
+#include "graph_interface.h"
+
+namespace vsag {
+class GraphInterfaceTest {
+public:
+    GraphInterfaceTest(std::shared_ptr<GraphInterface> graph) : graph_(graph){};
```
Comment (wxyucs) [inline src/data_cell/graph_interface_test.h]: use const reference of std::move

File: src/data_cell/sparse_graph_datacell_test.cpp [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+    auto max_id = 1000'000;
+    for (auto count : counts) {
+        auto graph = std::make_shared<SparseGraphDataCell>(graph_param, param);
+        GraphInterfaceTest test(graph);
+        auto other = std::make_shared<SparseGraphDataCell>(graph_param, param);
+        test.BasicTest(max_id, count, other);
+    }
+}
+
+TEST_CASE("graph basic test [ut][sparse graph datacell]") {
```
Comment (wxyucs) [inline src/data_cell/sparse_graph_datacell_test.cpp]: incorrect test case format

**Review by wxyucs — CHANGES_REQUESTED** [review]

File: src/data_cell/graph_datacell.h (hunk 1 above)
Comment (LHT129) [reply src/data_cell/graph_datacell.h]: done

File: src/data_cell/graph_datacell.h (hunk 2 above)
Comment (LHT129) [reply src/data_cell/graph_datacell.h:82]: https://github.com/LHT129/vsag/blob/graph_datacell/src/data_cell/graph_datacell.h#L98

File: src/data_cell/graph_datacell.h (hunk 1 above)
Comment (LHT129) [reply src/data_cell/graph_datacell.h]: done

File: src/data_cell/graph_datacell.h (hunk 4 above)
Comment (LHT129) [reply src/data_cell/graph_datacell.h]: done

File: src/data_cell/graph_datacell.h (hunk 3 above)
Comment (LHT129) [reply src/data_cell/graph_datacell.h]: truncate the vector by max_degree

File: src/data_cell/graph_datacell_test.cpp (hunk 5 above)
Comment (LHT129) [reply src/data_cell/graph_datacell_test.cpp]: done, but rename operator at other files may in another pull request

File: src/data_cell/graph_datacell_test.cpp (hunk 6 above)
Comment (LHT129) [reply src/data_cell/graph_datacell_test.cpp]: done

File: src/data_cell/graph_interface.cpp (hunk 7 above)
Comment (LHT129) [reply src/data_cell/graph_interface.cpp:2]: done

File: src/data_cell/graph_interface_test.cpp (hunk 9 above)
Comment (LHT129) [reply src/data_cell/graph_interface_test.cpp]: done

File: src/data_cell/graph_interface.cpp (hunk 8 above)
Comment (LHT129) [reply src/data_cell/graph_interface.cpp]: done

File: src/data_cell/graph_interface_test.h (hunk 10 above)
Comment (LHT129) [reply src/data_cell/graph_interface_test.h]: done

File: src/data_cell/sparse_graph_datacell.cpp (hunk 11 above)
Comment (LHT129) [reply src/data_cell/sparse_graph_datacell.cpp]: done

File: src/data_cell/sparse_graph_datacell.h (hunk 12 above)
Comment (LHT129) [reply src/data_cell/sparse_graph_datacell.h]: done

File: src/data_cell/sparse_graph_datacell.h (hunk 13 above)
Comment (LHT129) [reply src/data_cell/sparse_graph_datacell.h]: done

File: src/data_cell/sparse_graph_datacell.h (hunk 14 above)
Comment (LHT129) [reply src/data_cell/sparse_graph_datacell.h]: done

File: src/data_cell/sparse_graph_datacell.cpp (hunk 15 above)
Comment (LHT129) [reply src/data_cell/sparse_graph_datacell.cpp]: done

File: src/data_cell/sparse_graph_datacell_test.cpp (hunk 17 above)
Comment (LHT129) [reply src/data_cell/sparse_graph_datacell_test.cpp]: done

File: src/data_cell/graph_datacell.h (hunk 2 above)
Comment (wxyucs) [reply src/data_cell/graph_datacell.h:82]: add a const modifier

File: src/data_cell/sparse_graph_datacell.h [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    void
+    Serialize(StreamWriter& writer) override;
+
+    void
+    Deserialize(StreamReader& reader) override;
+
+private:
+    uint32_t code_line_size_{0};
```
Comment (wxyucs) [inline src/data_cell/sparse_graph_datacell.h:60]: add a const modifier

File: src/data_cell/graph_datacell.h (hunk 3 above)
Comment (wxyucs) [reply src/data_cell/graph_datacell.h]: add comments here about why logging error but not return early

**Review by wxyucs — CHANGES_REQUESTED** [review]

File: src/data_cell/graph_datacell.h (hunk 2 above)
Comment (LHT129) [reply src/data_cell/graph_datacell.h:82]: it can be changed if trans memory to disk

File: src/data_cell/graph_datacell.h (hunk 3 above)
Comment (LHT129) [reply src/data_cell/graph_datacell.h]: I think warning is better

File: src/data_cell/sparse_graph_datacell.h (hunk 18 above)
Comment (LHT129) [reply src/data_cell/sparse_graph_datacell.h:60]: the const for deserialize will be fixed in following pr

**Review by wxyucs — APPROVED** [review]
lgtm

File: src/data_cell/flatten_datacell.h [hunk 19]
```diff
@@ -30,8 +30,8 @@ class FlattenDataCell : public FlattenInterface {
 public:
     FlattenDataCell() = default;
 
-    explicit FlattenDataCell(const nlohmann::json& quantization_obj,
-                             const nlohmann::json& io_obj,
+    explicit FlattenDataCell(const JsonType& quantization_obj,
+                             const JsonType& io_obj,
```
Comment (jiaweizone) [inline src/data_cell/flatten_datacell.h]: `quantization_obj` and `io_obj` is configuration object ?
can named to `quantization_conf` / `io_conf` or `quantization_params` / `io_params`

File: src/data_cell/graph_datacell.h [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+    InsertNeighborsById(InnerIdType id, const Vector<InnerIdType>& neighbor_ids) override;
+
+    [[nodiscard]] uint32_t
+    GetNeighborSize(InnerIdType id) const override;
+
+    void
+    GetNeighbors(InnerIdType id, Vector<InnerIdType>& neighbor_ids) const override;
+
+    inline void
+    SetIO(std::shared_ptr<BasicIO<IOTmpl>> io) {
```
Comment (jiaweizone) [inline src/data_cell/graph_datacell.h:58]: minor: property accessor can named to `set_io(...)`

File: src/data_cell/graph_datacell.h [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+    Deserialize(StreamReader& reader) override;
+
+private:
+    std::shared_ptr<BasicIO<IOTmpl>> io_{nullptr};
+
+    uint32_t code_line_size_{0};
+};
+
+template <typename IOTmpl>
+GraphDataCell<IOTmpl, false>::GraphDataCell(const JsonType& graph_json_params,
```
Comment (jiaweizone) [inline src/data_cell/graph_datacell.h]: minor: `graph_json_params` and simply to `graph_params`

File: src/data_cell/graph_datacell.h [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    this->code_line_size_ = this->maximum_degree_ * sizeof(InnerIdType) + sizeof(uint32_t);
+}
+
+template <typename IOTmpl>
+void
+GraphDataCell<IOTmpl, false>::InsertNeighborsById(InnerIdType id,
+                                                  const Vector<InnerIdType>& neighbor_ids) {
+    if (neighbor_ids.size() > this->maximum_degree_) {
```
Comment (jiaweizone) [inline src/data_cell/graph_datacell.h:105]: The exceed `maximum_degree_` neighbor id will be discarded?

File: src/data_cell/graph_interface.cpp [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+#include "graph_interface.h"
+
+#include "graph_datacell.h"
+#include "io/io_headers.h"
+#include "sparse_graph_datacell.h"
+
+namespace vsag {
+
+GraphInterfacePtr
+GraphInterface::MakeInstance(const JsonType& json_obj,
```
Comment (jiaweizone) [inline src/data_cell/graph_interface.cpp]: json_obj -> build_conf / build_params ?

File: src/data_cell/graph_interface.cpp [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+
+    CHECK_ARGUMENT(json_obj.contains(IO_PARAMS_KEY),
+                   fmt::format("graph interface parameters must contains {}", IO_PARAMS_KEY));
+    const auto& io_param = json_obj[IO_PARAMS_KEY];
+
+    if (io_string == IO_TYPE_VALUE_BLOCK_MEMORY_IO) {
+        return std::make_shared<GraphDataCell<MemoryBlockIO, false>>(
+            graph_param, io_param, common_param);
+    }
+    if (io_string == IO_TYPE_VALUE_MEMORY_IO) {
```
Comment (jiaweizone) [inline src/data_cell/graph_interface.cpp]: minor: else if

File: src/data_cell/graph_interface.h [hunk 25]
```diff
@@ -24,32 +24,36 @@
...
-    GetNeighborSize(uint64_t id) = 0;
+    GetNeighborSize(InnerIdType id) const = 0;
 
     virtual void
-    GetNeighbors(uint64_t id, std::vector<uint64_t>& neighbor_ids) = 0;
+    GetNeighbors(InnerIdType id, Vector<InnerIdType>& neighbor_ids) const = 0;
 
     virtual void
-    Prefetch(uint64_t id, uint64_t neighbor_i) = 0;
+    Prefetch(InnerIdType id, uint32_t neighbor_i) = 0;
```
Comment (jiaweizone) [inline src/data_cell/graph_interface.h:56]: `neigbhor_id` change to `uint32_t` for big single node is enough ?

File: src/data_cell/sparse_graph_datacell.cpp [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "sparse_graph_datacell.h"
+
+namespace vsag {
+
+SparseGraphDataCell::SparseGraphDataCell(const JsonType& graph_json_params,
```
Comment (jiaweizone) [inline src/data_cell/sparse_graph_datacell.cpp]: simply to graph_params ?

File: src/data_cell/sparse_graph_datacell.cpp [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+                                         const IndexCommonParam& common_param)
+    : allocator_(common_param.allocator_), neighbors_(allocator_) {
```
Comment (jiaweizone) [inline src/data_cell/sparse_graph_datacell.cpp]: minor: Here depends the fields initialize sequence
can change to: neighbors_(common_param.allocator_)

File: src/data_cell/sparse_graph_datacell.cpp [hunk 28]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+SparseGraphDataCell::SparseGraphDataCell(Allocator* allocator, uint32_t max_degree)
+    : allocator_(allocator), neighbors_(allocator_) {
+    this->maximum_degree_ = max_degree;
+}
+
+void
+SparseGraphDataCell::InsertNeighborsById(InnerIdType id, const Vector<InnerIdType>& neighbor_ids) {
+    if (neighbor_ids.size() > this->maximum_degree_) {
```
Comment (jiaweizone) [inline src/data_cell/sparse_graph_datacell.cpp:39]: ditto

File: src/data_cell/flatten_datacell.h (hunk 19 above)
Comment (LHT129) [reply src/data_cell/flatten_datacell.h]: done

File: src/data_cell/graph_datacell.h (hunk 21 above)
Comment (LHT129) [reply src/data_cell/graph_datacell.h]: done

File: src/data_cell/graph_datacell.h (hunk 20 above)
Comment (LHT129) [reply src/data_cell/graph_datacell.h:58]: public method in upper case

File: src/data_cell/graph_datacell.h (hunk 22 above)
Comment (LHT129) [reply src/data_cell/graph_datacell.h:105]: yes

File: src/data_cell/graph_interface.cpp (hunk 23 above)
Comment (LHT129) [reply src/data_cell/graph_interface.cpp]: done

File: src/data_cell/graph_interface.cpp (hunk 24 above)
Comment (LHT129) [reply src/data_cell/graph_interface.cpp]: done

File: src/data_cell/graph_interface.h (hunk 25 above)
Comment (LHT129) [reply src/data_cell/graph_interface.h:56]: use InnerIdType (int32_t)

File: src/data_cell/sparse_graph_datacell.cpp (hunk 26 above)
Comment (LHT129) [reply src/data_cell/sparse_graph_datacell.cpp]: done

File: src/data_cell/graph_interface.h (hunk 25 above)
Comment (LHT129) [reply src/data_cell/graph_interface.h:56]: neighbor counts for one node, uint32_t is enough, The InnerIdType is uint32_t

File: src/data_cell/sparse_graph_datacell.cpp (hunk 27 above)
Comment (LHT129) [reply src/data_cell/sparse_graph_datacell.cpp]: done

File: src/data_cell/sparse_graph_datacell.cpp (hunk 28 above)
Comment (LHT129) [reply src/data_cell/sparse_graph_datacell.cpp:39]: ditto

**Pull request merged**